test/test_numpy.py

```diff
@@ -5,10 +5,18 @@
 
 import funsor
 from funsor import Number, Variable, bint, reals
+from funsor.interpreter import _USE_TCO
 from funsor.numpy import Array
 from funsor.testing import assert_equiv, check_funsor, random_array
 
 
+# FIXME rewrite stack-based interpreter to be compatible with unhashable data
```

Comment:
Do you plan to use the same pattern we used when cons-hashing?
https://github.com/pyro-ppl/funsor/blob/ba0dfcb4ff0fbe7a8a14866bf178a2ef2b290feb/funsor/terms.py#L83

Reply:
Yes, something like that.  Note that this only matters for numpy arrays; every other data term works fine.